src/devices_models/devices/common/cost_functions.jl

```diff
@@ -701,6 +749,86 @@ function add_to_cost!(
     return
 end
 
+function add_to_cost!(
```

Comment:
I don't get what you are trying to do here. You pass the spec and the cost_data simultaneously which in a way have the same information.